WMF Framework/Collection+AsyncMap.swift

```diff
@@ -1,37 +1,37 @@
 import Foundation
 
-extension Collection {
+public extension Collection {
     func asyncMap<R>(_ block: (Element, @escaping (R) -> Void) -> Void, completion:  @escaping ([R]) -> Void) {
         let group = DispatchGroup()
-        var results: [R] = []
-        for object in self {
+        var results = [R?](repeating: nil, count: count)
+        for (index, object) in self.enumerated() {
             group.enter()
             block(object, { result in
-                results.append(result)
+                results[index] = result
```

Comment:
A good use case would be anywhere you need to correlate the results to something. For example, fetching article summary objects from a list of titles, then updating the local articles with those summary results. I like the idea of a way to do this and get `[Element: R]` but probably would be better as a separate method since map corresponds to having a collection of n elements as input and getting a collection of n elements as output

Reply:
The docs for the [javascript version](https://www.w3schools.com/jsref/jsref_map.asp) has a pretty good definition of what's typically expected of a functional map method:

- The map() method creates a new array with the results of calling a function for every array element.

- The map() method calls the provided function once for each element in an array, in order.

- Note: map() does not execute the function for array elements without values.

- Note: map() does not change the original array.

Reply:
(though the JS one isn't async - you'd prob use Promise.all for that in JS land)